test/cluster/mzcompose.py

```diff
@@ -357,25 +357,18 @@ def workflow_test_github_15535(c: Composition) -> None:
     print("Sleeping to wait for frontier updates")
     time.sleep(10)
 
-    def extract_frontiers(output: str) -> Tuple[str, str]:
-        since_re = re.compile("^\s+since:\[(?P<frontier>.*)\]")
-        upper_re = re.compile("^\s+upper:\[(?P<frontier>.*)\]")
-        since = None
-        upper = None
-        for line in output.splitlines():
-            if match := since_re.match(line):
-                since = match.group("frontier").strip()
-            elif match := upper_re.match(line):
-                upper = match.group("frontier").strip()
-
-        assert since is not None, "since not found in EXPLAIN TIMESTAMP output"
-        assert upper is not None, "upper not found in EXPLAIN TIMESTAMP output"
-        return (since, upper)
+    def extract_frontiers(output: str) -> Tuple[int, int]:
+        j = json.loads(output)
+        upper_elements = j["determination"]["upper"]["elements"]
+        since_elements = j["determination"]["since"]["elements"]
+        assert len(upper_elements) == 1, "upper not found in EXPLAIN TIMESTAMP output"
+        assert len(since_elements) == 1, "since not found in EXPLAIN TIMESTAMP output"
```

Comment:
Nit: The assert message would be misleading if there would ever be more than one element.

What you could do to unpack the frontier into a single int:
```python
since, = since_elements
```

I think that should work and throw an exception if there are fewer or more than one element, but my Python is a bit rusty.

Reply:
Thanks, that seems nicer.